compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -12845,195 +12845,288 @@ TR::Node *su2dSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
-            && node->getNumChildren() > 0
-            && node->getFirstChild()->getOpCode().isInteger()))
-      return node;
-
-   TR::Node* opNode      = node->getFirstChild();
-   TR::Node* secondChild = node->getSecondChild();
-   TR::ILOpCodes op      = opNode->getOpCodeValue();
 
+   TR::ILOpCodes op = opNode->getOpCodeValue();
    // Get recognized operation types. Unsigned IL OpCodes are deprecated and not supported here.
    bool isAddOp = (op == TR::iadd || op == TR::ladd || op == TR::sadd || op == TR::badd);
-   bool isSubOp = (op == TR::isub || op == TR::lsub || op == TR::ssub || op == TR::bsub);
 
-   bool isUnsignedCompare = node->getOpCode().isUnsignedCompare();
-   bool isEqOrNotEqCompare = node->getOpCode().isCompareForEquality();
 
-   bool isSignednessTypeSupported = isEqOrNotEqCompare || (!isUnsignedCompare && opNode->cannotOverflow());
+   TR::DataType DataType = secondChild->getDataType();
+   if (!isSigned)
+      {
+      uint64_t oldUConst1 = 0, oldUConst2 = 0, unsignedMax = 0;
+      switch (DataType)
+         {
+         case TR::Int8:
+            oldUConst1 = opNode->getSecondChild()->getUnsignedByte();
+            oldUConst2 = secondChild->getUnsignedByte();
+            unsignedMax = TR::getMaxUnsigned<TR::Int8>();
+            break;
+         case TR::Int16:
+            oldUConst1 = opNode->getSecondChild()->getUnsignedShortInt();
+            oldUConst2 = secondChild->getUnsignedShortInt();
+            unsignedMax = TR::getMaxUnsigned<TR::Int16>();
+            break;
+         case TR::Int32:
+            oldUConst1 = opNode->getSecondChild()->getUnsignedInt();
+            oldUConst2 = secondChild->getUnsignedInt();
+            unsignedMax = TR::getMaxUnsigned<TR::Int32>();
+            break;
+         case TR::Int64:
+            oldUConst1 = opNode->getSecondChild()->getUnsignedLongInt();
+            oldUConst2 = secondChild->getUnsignedLongInt();
+            unsignedMax = TR::getMaxUnsigned<TR::Int64>();
+            break;
+         default:
+            TR_ASSERT_FATAL(false, "\nEliminating add/sub under compare node n%dn failed due to opcode data type\n",
+                            node->getGlobalIndex());
+            break;
+         }
+
 
-   if ((isAddOp || isSubOp)
-         && isSignednessTypeSupported
-         && opNode->getSecondChild()->getOpCode().isLoadConst()
-         && secondChild->getOpCode().isLoadConst()
-         // Only perform the simplification on the first occurance of the add/sub node
-         && (opNode->getFutureUseCount() == opNode->getReferenceCount() - 1))
+   // The unsigned case either only requires an overflow check (for a sub operation)
+   // or an underflow check (for an add operation). When one evaluates to true, the optimization can be performed.
+      if (!(isNotUnderflow(op, oldUConst1, oldUConst2,0)
+          || isNotOverflow(op, oldUConst1, oldUConst2, unsignedMax)))
```

Comment:
Uh, this isn't going to work. You can't use the signed functions to check for unsigned overflow/underflow.

Reply:
You're right. I had it that way originally, but changed it erroneously thinking that subOps for unsigned underflow would never reach this part in the code, and likewise in addOps for unsigned overflow

Reply:
Its been updated